proposals/public-suffix.md

```diff
@@ -0,0 +1,836 @@
+use cases are set out in the [Use Cases](#use-cases) section. Although the extension
+authors have not been consulted directly, it is assumed that they would stand to benefit
+from this API due to the reasons stated in the [Objective](#objective) section.
+
+| Extension | Relevant use cases | Description |
+|-----------|--------------------|-------------|
+| [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) | #1, #2 | isolate requests into containers by organizational base domain, view all domains grouped by organization |
+| [uBlock Origin](https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/) | #1, #2, #3 | create organization/3rd-party request-blocking rules, view request blocking status by organization, create blocking rules for specifically 3rd-party requests |
+| [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | #1, #3 | create organization/3rd-party request-blocking rules |
+| [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | #3 | exclude requests from tracking detection unless they are 3rd-party |
+| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any eTLD |
+
+## Specification
+
+### Schema
+
+A new API `publicSuffix` is added as follows:
+
+```ts
+namespace publicSuffix {
+
+
+  // METHODS
+
+  // Determines if the given hostname is itself a known eTLD (i.e. in the PSL).
+  export function isKnownSuffix(
+    hostname: string,
+  )
+  : boolean;
+
+  // Gets the known eTLD, if any, of a given hostname.
+  export function getKnownSuffix(
+    hostname: string,
+  )
+  : string | null;
+
+  // Gets the registrable domain of a given hostname.
+  export function getDomain(
+    hostname: string,
+    options?: DomainOptions,
+  )
+  : string | null;
+
+  // Gets the value of the VERSION metadata field in the PSL dataset if available
+  export function getVersion(): string | null;
+
+
+  // INTERFACES
+
+  // Options that may be passed to the API method to control its behaviour.
+  interface DomainOptions {
+
+    // If true, the returned domain should be encoded as Unicode.
+    // Default = false (Punycode)
+    unicode?: boolean,
+
+    // If true, and the input hostname is an IP address, then this is returned as-is.
+    // Default = false
+    allowIP?: boolean,
+
+    // If true, and the input hostname is itself a known eTLD (without a preceding label)
+    // then this is returned as-is.
+    // Default = false
+    allowPlainSuffix?: boolean,
+
+    // If true, and the input hostname lacks a known eTLD, and is neither itself
+    // a known eTLD nor an IP address, then the returned domain consists of
+    // the penultimate two domain labels of the input.
+    // Default = false
+    allowUnknownSuffix?: boolean,
+  }
```

Comment:
Should we provide an option for whether to include private registries?  (I think in this proposal, called "implied suffixes"?)

Reply:
Are you referring to the "Private section" of the PSL dataset? If so, these private registries are always included, as explained in proposal section "Features of the PSL: 1. ICANN vs Private".